app/workers/delete_object_hierarchy_worker.rb

```diff
@@ -73,49 +74,71 @@ def batch_callbacks(batch)
     end
   end
 
-  def delete_associations
-    destroyable_associations.each do |association|
-      ReflectionDestroyer.new(object, association, caller_worker_hierarchy).destroy_later
+  def destroy_and_delete_associations
+    Array(object.background_deletion).each do |association, options|
+      ReflectionDestroyer.new(object, association, options, caller_worker_hierarchy).destroy_later if destroyable_association?(association)
     end
   end
 
-  private
-
-  def destroy_method
-    @destroy_method.presence || 'destroy'
+  def destroyable_association?(_association)
+    true
   end
 
+  private
+
   def called_from_provider_hierarchy?
     return unless (tenant_id = object.tenant_id)
     caller_worker_hierarchy.include?("Hierarchy-Account-#{tenant_id}")
   end
 
-  def destroyable_associations
-    object.class.reflect_on_all_associations.select { |reflection| destroyable_association?(reflection) }
-  end
-
-  def destroyable_association?(reflection)
-    %i[destroy delete_all].include?(reflection.options[:dependent])
-  end
-
   def callback_options
     { 'object_global_id' => object.to_global_id, 'caller_worker_hierarchy' => caller_worker_hierarchy }
   end
 
   class ReflectionDestroyer
 
-    def initialize(main_object, reflection, caller_worker_hierarchy)
+    class FakeReflection
+      attr_reader :name, :options
+
+      def initialize(name, options)
+        @name = name
+        @options = options
+      end
+
+      def many?
+        return options[:has_many] if options.keys.include?(:has_many)
+
+        plural?
+      end
+
+      def class_name
+        options[:class_name].presence || name.to_s.singularize.classify
+      end
+
+      def background_destroy_method
+        options[:action].to_s
+      end
+
+      private
+
+      def plural?
+        string_name = name.to_s
```

Comment:
Well if name is `:users` then `:users.to_s.pluralize == "users"` that works fine

BUT invariable word will not:

`"information".pluralize == "information"`

Reply:
I agree with Hery that this is playing with fire, for example for 'Settings' 🤔 
I'd go for enforce the option 'has_many/has_one' or just 'one/many' 😄

Reply:
it's not bulletproof but it does work for 95% of cases.. for that 5 % is going to fail (TDD) and then you can specify `has_many` option.

Reply:
it's either configure everything manually or https://github.com/3scale/porta/pull/1372#discussion_r348069440

Reply:
i have an idea how to prevent `information. pluralize ` problem..